test_conformance/subgroups/test_subgroup_ballot.cpp

```diff
@@ -0,0 +1,1089 @@
+                        wi_id, operation_names(operation), max_sub_group_size);
+                    set_value(t[wg_offset + wi_id], expected_mask);
+                }
+            }
+
+            // Now map into work group using map from device
+            for (wi_id = 0; wi_id < lws; ++wi_id)
+            {
+                x[wi_id] = t[wi_id];
+            }
+            x += lws;
+            m += 4 * lws;
+        }
+    }
+
+    static int chk(Ty *x, Ty *y, Ty *mx, Ty *my, cl_int *m,
+                   const WorkGroupParams &test_params)
+    {
+        int wi_id, wg_id, sb_id;
+        int gws = test_params.global_workgroup_size;
+        int lws = test_params.local_workgroup_size;
+        int sbs = test_params.subgroup_size;
+        int sb_number = (lws + sbs - 1) / sbs;
+        Ty expected_result, device_result;
+        int wg_number = gws / lws;
+
+        for (wg_id = 0; wg_id < wg_number; ++wg_id)
+        { // for each work_group
+            for (wi_id = 0; wi_id < lws; ++wi_id)
+            { // inside the work_group
+                mx[wi_id] = x[wi_id]; // read host inputs for work_group
+                my[wi_id] = y[wi_id]; // read device outputs for work_group
+            }
+
+            for (sb_id = 0; sb_id < sb_number; ++sb_id)
+            {
+                int wg_offset = sb_id * sbs;
+                int current_sbs = wg_offset + sbs > lws ? lws - wg_offset : sbs;
+
+                // Check result
+                for (wi_id = 0; wi_id < current_sbs; ++wi_id)
+                { // inside the subgroup
+                    expected_result =
+                        mx[wg_offset + wi_id]; // read host input for subgroup
+                    device_result =
+                        my[wg_offset
+                           + wi_id]; // read device outputs for subgroup
+                    if (!compare(device_result, expected_result))
+                    {
+                        log_error("ERROR:  get_sub_group_%s_mask... mismatch "
+                                  "for local id %d in sub group %d in group "
+                                  "%d, obtained %d, expected %d\n",
+                                  operation_names(operation), wi_id, sb_id,
+                                  wg_id, device_result, expected_result);
+                        return TEST_FAIL;
+                    }
+                }
+            }
+            x += lws;
+            y += lws;
+            m += 4 * lws;
+        }
+        log_info("  get_sub_group_%s_mask... passed\n",
+                 operation_names(operation));
+        return TEST_PASS;
+    }
+};
+
+static const char *bcast_non_uniform_source =
+    "__kernel void test_bcast_non_uniform(const __global Type *in, __global "
+    "int4 *xy, __global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    Type x = in[gid];\n"
+    "    if (xy[gid].x < NR_OF_ACTIVE_WORK_ITEMS) {\n"
```

Comment:
Please rewrite the testing of non-uniformity here to match that for cl_khr_subgroup_non_uniform_vote with changes to reasonably ensure correctness.

This comment applies to all kernels that test "< NR_OF_ACTIVE_WORK_ITEMS". This is not an adequate level of testing for any of the extended subgroup functions.

Reply:
This issue is addressed by #1198

Reply:
Agreed.